fgpyo/sam/builder.py

```diff
@@ -503,34 +503,49 @@ def to_path(
         path: Optional[Path] = None,
         index: bool = True,
         pred: Callable[[AlignedSegment], bool] = lambda r: True,
+        tmp_file_type: Optional[sam.SamFileType] = None,
     ) -> Path:
         """Write the accumulated records to a file, sorts & indexes it, and returns the Path.
         If a path is provided, it will be written to, otherwise a temporary file is created
         and returned.
 
+        If `path` is provided, `tmp_file_type` may not be provided. In this case, the file type
+        (SAM/BAM/CRAM) will be automatically determined by the file extension when a path
+        is provided.  See `~pysam` for more details.
+
+        If `path` is not provided, the file type will default to BAM unless `tmp_file_type` is
+        provided.
+
         Args:
             path: a path at which to write the file, otherwise a temp file is used.
-            index: if True and `sort_order` is `Coordinate` index is generated, otherwise not.
+            index: if True and `sort_order` is `Coordinate` and output is a BAM file, then
+                   an index is generated, otherwise not.
             pred: optional predicate to specify which reads should be output
+            tmp_file_type: the file type to output when a path is not provided (default is BAM)
 
         Returns:
             Path: The path to the sorted (and possibly indexed) file.
         """
+        if path is not None and tmp_file_type is not None:
+            raise ValueError("Both `path` and `tmp_file_type` cannot be provided.")
+
+        if tmp_file_type is None:
+            tmp_file_type = sam.SamFileType.BAM
+
+        ext = tmp_file_type.extension
 
         if path is None:
-            with NamedTemporaryFile(suffix=".bam", delete=False) as fp:
+            with NamedTemporaryFile(suffix=ext, delete=False) as fp:
                 path = Path(fp.name)
 
-        with NamedTemporaryFile(suffix=".bam", delete=True) as fp:
+        with NamedTemporaryFile(suffix=ext, delete=True) as fp:
```

Comment:
All the if-elses and context managers are making this tricky for me to logic through. I'm still wondering if this works correctly for all cases. It looks like _**if**_ a path is defined, then the extension (`ext`) is always going to be BAM. However, if the sort order is **_neither_** queryname or coordinate, there is no pysam sort to the final file (by filename, not by open file handle).

So, for this combination of values, do you indeed get a named SAM file?

- Path suffixed with `.sam`
- Sort order _**neither**_ queryname or coordinate
- Leaving `tmp_file_type` as None

Reply:
I see code rabbit spotted the same issue. I testify I saw it first despite being late to hit "submit review" ;)